### ISSUE TITLE
[R-package] Same code in different machine result in different

### ISSUE THREAD

**Issue by zzhou93** [microsoft/LightGBM#6303]
## Description
Run the same code in different machines, Similar lgb.train results, but models' accuracy is pretty different.

## Reproducible example

```shell
Train_percent<-cdl_percent[,114:163]
colnames(Train_percent)
train_data <- as.matrix(Train_percent)
label <- as.numeric(as.factor(cdl_percent$BU_combined))-1
table(label)
train_dataset <- lgb.Dataset(data = train_data, label = label)
# Set parameters for LightGBM
params <- list(
  objective = "multiclass",
  metric = "multi_logloss",
  num_class = 8,
  learning_rate = 0.1,
  num_leaves = 50,
  num_threads = 8
)

# Train the LightGBM model
model <- lgb.train(
  params = params,
  data = train_dataset,
  nrounds = 10
)

new_data <- as.matrix(Train_percent)
predictions <- matrix(predict(model, new_data),nrow = nrow(Train_percent),byrow = T)

pred_y = apply(predictions,1,which.max)

as.numeric(summary(pred_y==label+1))[3]/length(pred_y)

caret::confusionMatrix(data = as.factor(BU_name[pred_y]), reference = as.factor(BU_name[(label+1)]))

```

## Environment info

LightGBM version or commit hash:
Machine 1：
R version 4.3.2 (2023-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 11 x64 (build 22631)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils    
[5] datasets  methods   base     

other attached packages:
 [1] lightgbm_3.3.5    R6_2.5.1         
 [3] lubridate_1.9.2   forcats_1.0.0    
 [5] stringr_1.5.0     dplyr_1.1.2      
 [7] purrr_1.0.1       readr_2.1.4      
 [9] tidyr_1.3.0       tibble_3.2.1     
[11] ggplot2_3.4.2     tidyverse_2.0.0  
[13] markovchain_0.9.3

loaded via a namespace (and not attached):
 [1] gtable_0.3.3        
 [2] xfun_0.39           
 [3] recipes_1.0.7       
 [4] lattice_0.21-9      
 [5] tzdb_0.4.0          
 [6] vctrs_0.6.2         
 [7] tools_4.3.2         
 [8] generics_0.1.3      
 [9] stats4_4.3.2        
[10] parallel_4.3.2      
[11] fansi_1.0.4         
[12] pkgconfig_2.0.3     
[13] ModelMetrics_1.2.2.2
[14] Matrix_1.6-1.1      
[15] data.table_1.14.8   
[16] RcppParallel_5.1.7  
[17] lifecycle_1.0.3     
[18] compiler_4.3.2      
[19] munsell_0.5.0       
[20] codetools_0.2-19    
[21] htmltools_0.5.5     
[22] class_7.3-22        
[23] yaml_2.3.7          
[24] prodlim_2023.03.31  
[25] pillar_1.9.0        
[26] MASS_7.3-60         
[27] gower_1.0.1         
[28] iterators_1.0.14    
[29] rpart_4.1.21        
[30] foreach_1.5.2       
[31] nlme_3.1-163        
[32] parallelly_1.36.0   
[33] lava_1.7.2.1        
[34] tidyselect_1.2.0    
[35] digest_0.6.31       
[36] stringi_1.7.12      
[37] future_1.33.0       
[38] reshape2_1.4.4      
[39] listenv_0.9.0       
[40] splines_4.3.2       
[41] fastmap_1.1.1       
[42] grid_4.3.2          
[43] colorspace_2.1-0    
[44] expm_0.999-7        
[45] cli_3.6.1           
[46] magrittr_2.0.3      
[47] survival_3.5-7      
[48] utf8_1.2.3          
[49] future.apply_1.11.0 
[50] withr_2.5.0         
[51] scales_1.2.1        
[52] timechange_0.2.0    
[53] rmarkdown_2.22      
[54] globals_0.16.2      
[55] igraph_1.5.0        
[56] nnet_7.3-19         
[57] timeDate_4022.108   
[58] hms_1.1.3           
[59] evaluate_0.21       
[60] knitr_1.43          
[61] hardhat_1.3.0       
[62] caret_6.0-94        
[63] rlang_1.1.1         
[64] Rcpp_1.0.10         
[65] glue_1.6.2          
[66] pROC_1.18.4         
[67] ipred_0.9-14        
[68] rstudioapi_0.14     
[69] jsonlite_1.8.5      
[70] plyr_1.8.8     

Machine2: 
R version 4.3.2 (2023-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods  
[7] base     

other attached packages:
[1] lightgbm_4.3.0

loaded via a namespace (and not attached):
 [1] utf8_1.2.4           future_1.33.1        generics_0.1.3      
 [4] class_7.3-22         stringi_1.8.3        lattice_0.21-9      
 [7] pROC_1.18.5          listenv_0.9.1        digest_0.6.34       
[10] magrittr_2.0.3       evaluate_0.23        grid_4.3.2          
[13] timechange_0.3.0     iterators_1.0.14     jsonlite_1.8.8      
[16] foreach_1.5.2        plyr_1.8.9           Matrix_1.6-1.1      
[19] ModelMetrics_1.2.2.2 nnet_7.3-19          survival_3.5-7      
[22] purrr_1.0.2          fansi_1.0.6          scales_1.3.0        
[25] codetools_0.2-19     lava_1.7.3           cli_3.6.2           
[28] rlang_1.1.3          caret_6.0-94         hardhat_1.3.1       
[31] parallelly_1.36.0    future.apply_1.11.1  munsell_0.5.0       
[34] splines_4.3.2        withr_3.0.0          prodlim_2023.08.28  
[37] tools_4.3.2          parallel_4.3.2       reshape2_1.4.4      
[40] dplyr_1.1.4          colorspace_2.1-0     recipes_1.0.9       
[43] ggplot2_3.4.4        globals_0.16.2       vctrs_0.6.5         
[46] R6_2.5.1             rpart_4.1.21         stats4_4.3.2        
[49] lifecycle_1.0.4      lubridate_1.9.3      stringr_1.5.1       
[52] MASS_7.3-60          pkgconfig_2.0.3      pillar_1.9.0        
[55] gtable_0.3.4         glue_1.7.0           data.table_1.15.0   
[58] Rcpp_1.0.12          highr_0.10           xfun_0.41           
[61] tibble_3.2.1         tidyselect_1.2.0     knitr_1.45          
[64] rstudioapi_0.15.0    nlme_3.1-163         ipred_0.9-14        
[67] timeDate_4032.109    gower_1.0.1          compiler_4.3.2  

Command(s) you used to install LightGBM

## Additional Comments

With Machine 1 accuracy is 84% and 2 accuracy is 11.5%.
![Screenshot 2024-02-06 183805](https://github.com/microsoft/LightGBM/assets/42654457/11ca79a9-b982-41b3-bd92-70adb50654c3)
![Screenshot 2024-02-06 183932](https://github.com/microsoft/LightGBM/assets/42654457/f4be4f22-e26b-477b-a457-fd14919f60b9)

**Comment by jmoralez** [comment]
Hey @zzhou93, thanks for using LightGBM. I see that you have LightGBM 3.3.5 and 4.3.0 respectively. Can you please use the same version in both? There was a recent change on how the multiclass predictions are returned so that may be the source of the difference.

**Comment by zzhou93** [comment]
Yes. Thank you very much. The version 3.3.5 works well. But could you provide me more details about the recent change that affects the multiclass predictions

**Comment by zzhou93** [comment]
Thank you so much, the problem solved.
